Add Radio class abstracting  otPlatRadio APIs

abtink opened the pull request:
This commit contains two related commits:

**[radio] add Radio class abstracting radio platform APIs**
    
 This commit adds a a `Radio` class which provides all `otPlatRadio`
 platform APIs (as C++ methods). All OT core modules (except for diags)
 use the new `Radio` methods to interact with radio platform APIs.

---------------

**[radio] add radio folder to core**
    
 This commit renames existing `phy` folder into `radio`. It also
 renames some of the related files in this folder.

------------

The objective behind adding the `Radio` class is to provide a layer where we can add support for alternate (internal) radio implementations (e.g. Thread over BLE (ToBLE) abstraction) or multi-radio.

jwhui commented on src/core/radio/radio.hpp:
```diff
@@ -0,0 +1,417 @@
+    /**
+     * This method gets the radio receive sensitivity value.
+     *
+     * @param[in] aInstance  The OpenThread instance structure.
```
Remove `@param`.

jwhui commented on src/core/radio/radio.hpp:
```diff
@@ -0,0 +1,417 @@
+    /**
+     * This method gets the status of promiscuous mode.
+     *
+     *
```
nit: extra line

jwhui commented on src/core/radio/radio.hpp:
```diff
@@ -0,0 +1,417 @@
+     *
+     * OpenThread forms the IEEE 802.15.4 frame in this buffer then calls `Transmit()` to request transmission.
+     *
+     *
```
nit: extra line

jwhui commented on src/core/radio/radio.hpp:
```diff
@@ -0,0 +1,417 @@
+    /**
+     * This method adds a short address to the source address match table.
+     *
+     * @param[in]  aInstance      The OpenThread instance structure.
```
Remove `aInstance`

jwhui commented on src/core/radio/radio.hpp:
```diff
@@ -0,0 +1,417 @@
+ * This class represents an OpenThread radio abstraction.
+ *
+ */
+class Radio : public InstanceLocator
```
Should `otPlatRadio*` callbacks go through this class as well?

abtink replied:
Was actually thinking of the same thing. :)

turon commented on src/core/common/instance.hpp:
```diff
@@ -312,17 +312,23 @@ class Instance : public otInstance
-#if !OPENTHREAD_CONFIG_MULTIPLE_INSTANCE_ENABLE
+
+#if OPENTHREAD_MTD || OPENTHREAD_FTD
+#if !OPENTHREAD_ENABLE_MULTIPLE_INSTANCES
```
Is this intended to regress the recent move to `OPENTHREAD_CONFIG_MULTIPLE_INSTANCE_ENABLE` above?

abtink replied:
The diff here is a bit hard to read. The change is actually just moving the comment above the `#if` checks instead of in middle of them.

jwhui commented:
Should the following also make use of the new `Radio` class?
1. `ot::FactoryDiags::Diags` 
1. `otGetVersionString()` - I realize this function does not take `aInstance` as a parameter - should it?

turon commented on src/core/Makefile.am:
```diff
@@ -194,7 +194,7 @@ SOURCES_COMMON                             = \
     net/sntp_client.cpp                      \
     net/udp6.cpp                             \
-    phy/radio_weak.cpp                       \
+    radio/radio_platform_weak.cpp            \
```
Perhaps rename this file `radio_stub.cpp`?

abtink commented:
Pushed a new set of commits. 
- Rebased on master.
- Changed diags to also use the `Radio` APIs.
- Added a new commit to add `Radio::Callbacks` to provide the platform radio callbacks. 

---------------------

**[radio] adding Radio::Callback class**

This commit adds a `Radio::Callback` class which provides all the
callbacks (as C++ methods) mirroring the radio platform layer
callbacks (including the diags related radio callbacks).

jwhui commented on src/core/radio/radio.hpp:
```diff
@@ -0,0 +1,518 @@
+         * @param[in]  aError     OT_ERROR_NONE when the frame was transmitted,
+         *                        OT_ERROR_NO_ACK when the frame was transmitted but no ACK was received,
+         *                        OT_ERROR_CHANNEL_ACCESS_FAILURE tx could not take place due to activity on the
+         * channel, OT_ERROR_ABORT when transmission was aborted for other reasons.
```
nit: alignment

jwhui commented on src/core/radio/radio.hpp:
```diff
@@ -0,0 +1,518 @@
+         * @param[in]  aError     OT_ERROR_NONE when the frame was transmitted,
+         *                        OT_ERROR_NO_ACK when the frame was transmitted but no ACK was received,
+         *                        OT_ERROR_CHANNEL_ACCESS_FAILURE tx could not take place due to activity on the
+         * channel, OT_ERROR_ABORT when transmission was aborted for other reasons.
```
nit: alignment
